src/controlsocket.py

```diff
@@ -0,0 +1,227 @@
+        to the URL. If body dict is provided, it is serialied as JSON and used
+        as the HTTP body (with Content-Type: "application/json"). The resulting
+        body is decoded from JSON.
+        """
+        headers = {'Accept': 'application/json'}
+        data = None
+        if body is not None:
+            data = json.dumps(body).encode('utf-8')
+            headers['Content-Type'] = 'application/json'
+
+        response = self._request_raw(method, path, query, headers, data)
+        self._ensure_content_type(response.headers, 'application/json')
+        raw_resp: Dict[str, Any] = json.loads(response.read())
+        return raw_resp
+
+    @staticmethod
+    def _ensure_content_type(headers: email.message.Message,
+                             expected: 'Literal["multipart/form-data", "application/json"]'):
+        """Parse Content-Type header from headers and ensure it's equal to expected.
+
+        Return a dict of any options in the header, e.g., {'boundary': ...}.
+        """
+        ctype = headers.get_content_type()
+        params = headers.get_params() or {}
+        options = {key: value for key, value in params if value}
+        if ctype != expected:
+            raise ProtocolError(f'expected Content-Type {expected!r}, got {ctype!r}')
+        return options
+
+    def _request_raw(
+        self, method: str, path: str,
+        query: Optional[Dict[str, Any]] = None,
+        headers: Optional[Dict[str, Any]] = None,
+        data: Optional[Union[bytes, Generator[bytes, Any, Any]]] = None,
+    ) -> http.client.HTTPResponse:
+        """Make a request to the socket; return the raw HTTPResponse object."""
+        url = self.base_url + path
+        if query:
+            url = f"{url}?{urllib.parse.urlencode(query, doseq=True)}"
+
+        if headers is None:
+            headers = {}
+        request = urllib.request.Request(url, method=method, data=data, headers=headers)
+
+        try:
+            response = self.opener.open(request, timeout=self.timeout)
+        except urllib.error.HTTPError as e:
+            code = e.code
+            status = e.reason
+            try:
+                body: Dict[str, Any] = json.loads(e.read())
+                message: str = body['result']['message']
+            except (OSError, ValueError, KeyError) as e2:
+                # Will only happen on read error or if the server sends invalid JSON.
+                body: Dict[str, Any] = {}
+                message = f'{type(e2).__name__} - {e2}'
+            raise APIError(body, code, status, message)
+        except urllib.error.URLError as e:
+            raise ConnectionError(e.reason)
+
+        return response
+    
+    def add_metrics_user(self, username: str, password: str):
+        resp = self._request(
+            method='POST',
+            path='/metrics-users',
+            body={"username": username, "password": password},
+        )
+        logger.info(f'result of add metrics user request: {resp}')
```

Comment:
Should these be `logger.debug`? It seems to me if the request is successful we don't want an INFO log for each one.